### ISSUE TITLE
What is the purpose of a predicate?

### ISSUE THREAD

**Issue by soluchok** [decentralized-identity/presentation-exchange#173]
`predicate` is not described clearly (at least for me). It would be nice to have an example of what predicate does.
Here is an example of how I understand it.
https://github.com/hyperledger/aries-framework-go/tree/master/pkg/doc/presexch
Can someone confirm whether it is correct?

**Comment by brentzundel** [comment]
Your example looks correct to me.

**Comment by llorllale** [comment]
Related: #46, #85.

@brentzundel (and anyone else following this):

This ticket was raised due to different interpretations of the spec between @soluchok and myself regarding how `predicate` works.

I will replicate the samples from the document @soluchok linked to above for reference:

<details><summary>Presentation Definition: request age equal or greater than 18 with predicate</summary>

```json
{
  "id": "31e2f0f1-6b70-411d-b239-56aed5321884",
  "purpose": "To sell you a drink we need to know that you are an adult.",
  "input_descriptors": [
    {
      "id": "867bfe7a-5b91-46b2-9ba4-70028b8d9cc8",
      "purpose": "Your age should be greater or equal to 18.",
      "schema": [
        {
          "uri": "https://www.w3.org/TR/vc-data-model/#types"
        }
      ],
      "constraints": {
        "limit_disclosure": true,
        "fields": [
          {
            "path": [
              "$.age"
            ],
            "filter": {
              "type": "integer",
              "minimum": 18
            },
            "predicate": "required"
          }
        ]
      }
    }
  ]
}
```
</details>

<details><summary>Candidate Input: age = 18</summary>

```json
{
  "@context": [
    "https://www.w3.org/2018/credentials/v1"
  ],
  "age": 21,
  "credentialSchema": [
    {
      "id": "https://www.w3.org/TR/vc-data-model/#types"
    }
  ],
  "first_name": "Jesse",
  "id": "2dc74354-e965-4883-be5e-bfec48bf60c7",
  "issuer": "",
  "last_name": "Pinkman",
  "type": "VerifiableCredential"
}
```
</details>

<details><summary>Presentation Definition: credential with age=true</summary>

```json
{
  "@context": [
    "https://www.w3.org/2018/credentials/v1",
    "https://identity.foundation/presentation-exchange/submission/v1"
  ],
  "presentation_submission": {
    "id": "accd5adf-1dbf-4ed9-9ba2-d687476126cb",
    "definition_id": "31e2f0f1-6b70-411d-b239-56aed5321884",
    "descriptor_map": [
      {
        "id": "867bfe7a-5b91-46b2-9ba4-70028b8d9cc8",
        "format": "ldp_vp",
        "path": "$.verifiableCredential[0]"
      }
    ]
  },
  "type": [
    "VerifiablePresentation",
    "PresentationSubmission"
  ],
  "verifiableCredential": [
    {
      "@context": [
        "https://www.w3.org/2018/credentials/v1"
      ],
      "age": true,
      "credentialSchema": [
        {
          "id": "https://www.w3.org/TR/vc-data-model/#types"
        }
      ],
      "credentialSubject": null,
      "id": "2dc74354-e965-4883-be5e-bfec48bf60c7",
      "issuer": "",
      "type": "VerifiableCredential"
    }
  ]
}
```
</details>

As I read section `5.3` initially I believed the following to be merely part of the description of an algorithm on how a Holder ("processing entity") must evaluate a PE:

> If the predicate property of the field entry is present, a boolean value should be returned rather than the value of the Field Query Result. Calculate this boolean value by evaluating the Field Query Result against the JSON Schema descriptor value of the filter property.
 
But then I read this paragraph in section `5.2.1` and I start to get confused:

> If the predicate property is present, the filter property MUST also be present. The inclusion of the predicate property indicates that the processing entity returns a boolean, rather than a value returned from evaluation of the JSONPath string expressions in the path array. The boolean returned is the result of using the filter property’s JSON Schema descriptors against the evaluated value. 

So it _does_ look like the intent of `predicate` is to transform the original claim into a boolean for the Verifier. The boolean value will always be `true` because of the pairing with `filter`. This is what is depicted in the examples above.

This has a few issues:

* This effectively limits its use in practice to the Indy ZKP mechanisms (or any other like it that I'm not aware of). BBS+ is not suitable for this. Any implementer not using Indy-like ZKPs must either not engage in the presentation exchange, or attempt real-time issuance of a credential that conforms to the `filter`. **Are there any ZKP schemes out there incubated in a standards body that we could use for this?**
* It seems to contradict - at least in part - the idea that credentials requestd with P-E have their structures defined with JSONSchema, and that the P-E `schema` essentially indicates the credential structure the Verifier is requesting. `schema` and `filter` play together nicely in this regard: the Holder indexes their credentials by their schema URI and then runs the JSONSchema filters on them to find matches. The Verifier can verify the credential's structure against the same JSONSchema. But `predicate` is instructing the Holder to output a credential in a different schema than the one they are currently holding. This complicates things for the Holder (with a potentially costly performance penalty - see previous point) as well as for the Verifier, how now needs to run the credential through a different predicate to determine whether it satisfies the requirements.

**Comment by llorllale** [comment]
Another problem with this:

Selective disclosure with BBS+ relies on reusing the original claims in the credential. There is a big expectation in the Aries community to use P-E with BBS+ for selective disclosure capabilities.

`predicate=required` breaks support for selective disclosure with BBS+ as the credential being verified has claims that were not originally signed over by the Issuer.

**Comment by TomCJones** [comment]
at the meeting this morning 1-14, it was asserted that a predicate must be true or false.  That statement is false. There are two cases where such a determination cannot be made. That would return "null" in most programming languages.
1. the statement is not a well-formed-formula
2. the statement is not able to provide semantic meaning by means of an interpretation
An example would be "is-over-21" only makes sense if the current date and birth date are both available.

**Comment by brentzundel** [comment]
I will try to address the questions above.

> This effectively limits its use in practice to the Indy ZKP mechanisms (or any other like it that I'm not aware of).

This may be true.

Any use of the `predicate` property requires transformation from the attribute value in the original credential into a predicate value derived credential in order to be presented. Other predicate mechanisms in which issuers add predicate values to the issued credential can be used like any other attribute and a verifier does not need to use the `predicate` property to request them.

> Any implementer not using Indy-like ZKPs must either not engage in the presentation exchange, or attempt real-time issuance of a credential that conforms to the filter.

The `predicate` property is not required. A verifier who wishes to request predicate values must understand the limitations of the signature schemes they support. It is possible for a verifier to state they only support JWTs and to have a `"predicate": "required"` flag, which would be impossible to comply with (at this point I am not aware of any predicate-supportive signatures that work with JWTs).
A design decision we made pretty early in developing this specification was that we are not going to prevent valid use cases in order to protect a verifier from being able to create an invalid presentation definition, but confusion around this point may mean that additional text to clarify predicates should be added to the spec.

> Are there any ZKP schemes out there incubated in a standards body that we could use for this?

To my knowledge, there are no ZKP schemes incubated by a standards body which support predicates. Should this be a requirement for the spec? Should we require that any mechanism which enables compliant presentations be described in a standard somewhere? How could we enforce that?

> It seems to contradict - at least in part - the idea that credentials requestd with P-E have their structures defined with JSONSchema, and that the P-E schema essentially indicates the credential structure the Verifier is requesting. schema and filter play together nicely in this regard: the Holder indexes their credentials by their schema URI and then runs the JSONSchema filters on them to find matches. The Verifier can verify the credential's structure against the same JSONSchema. But predicate is instructing the Holder to output a credential in a different schema than the one they are currently holding. This complicates things for the Holder (with a potentially costly performance penalty - see previous point) as well as for the Verifier, how now needs to run the credential through a different predicate to determine whether it satisfies the requirements.

As you express, holder processing and verifier processing are separate mechanisms. It is my understanding that the schema URI and JSONSchema filters in the presentation definition are provided to enable holder processing of potential credentials.
The presentation definition itself serves as the schema which describes the presentation. The verifier would use the presentation manifest to evaluate the presentation's conformance to the presentation definition and to process the presentation.

The verifiable credentials specification clearly indicates that a submitted presentation may be made up of derived credentials rather than original credentials, and that these derived credentials may not comply with the schema that describes the credential.

While predicates certainly make processing trickier, they are only one capability in the spec that does so. Some selective disclosure mechanisms also result in derived credentials that would not match the credential schema included with the presentation definition. 

> Selective disclosure with BBS+ relies on reusing the original claims in the credential. There is a big expectation in the Aries community to use P-E with BBS+ for selective disclosure capabilities.
> predicate=required breaks support for selective disclosure with BBS+ as the credential being verified has claims that were not originally signed over by the Issuer.

BBS+ signatures do not natively support predicates. Nor do LD-Signatures, nor any other digital signature scheme that isn't ZKP-capable. A verifier who wishes to request selectively disclosed values and predicates, while supporting the widest variety of signatures, should be advised to use `predicate=preferred`. 

tl;dr - Additional clarifying language and usage guidance for predicates should be added to the spec. I will raise a PR to do so.

**Comment by brentzundel** [comment]
> at the meeting this morning 1-14, it was asserted that a predicate must be true or false. That statement is false. There are two cases where such a determination cannot be made. That would return "null" in most programming languages.
> 
> 1. the statement is not a well-formed-formula
> 2. the statement is not able to provide semantic meaning by means of an interpretation
>    An example would be "is-over-21" only makes sense if the current date and birth date are both available.

iiuc, the first case would be an example of an invalid presentation definition. We understand that it is nearly impossible to have a functional spec and a spec that prevents invalid presentation definitions.

the second case would be an example of 1) the verifier not providing the proper JSONSchema filter (i.e., the current date is not provided, resulting in an invalid presentation definition), or 2) the holder not having a credential which matches the provided JSON Path results (i.e., the birth date).

@TomCJones it seems like the two cases you highlight constitute either invalid presentation definitions, or the holder doesn't have the requested data, neither of which are cases unique to predicates.
I would consider both of these to be categorically different than predicates resulting in true or false values, and therefore not be considered as possible results from performing a predicate operation, but rather as possible results from holder processing of the presentation definition itself.

**Comment by csuwildcat** [comment]
This Issue was resolved on the latest PE call. More specific discussion/actions related to this general topic can be found here: https://github.com/decentralized-identity/presentation-exchange/issues/175
